### DISCUSSION TITLE
Nothing rendered in Browser

### DISCUSSION THREAD

**Discussion by jamesonuk** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#439] [Q&A]
I have set this up on 1 D1 mini (https://smile.amazon.co.uk/AZDelivery-Development-ESP8266EX-Compatible-Micropython/dp/B08BTYHJM1/) as per https://antanaitis.lt/2022/09/07/wi-fi-on-your-lay-z-spa-make-it-easy-to-remote-control/

This seems to flash fine, I can connect and set the WiFi credentials but then things start to go awry

Straight after flashing and updating the WiFi details I can normally get to the home page and sometimes open up some of the child pages but at some point very quickly after that it just stops rendering

I have just gone in now (I flashed it this morning) and if I connect a serial monitor I see
```
HTTP > request: /
HTTP > file sent: /index.html (11023 bytes)
HTTP > request: /WebSocket.js
HTTP > file sent: /WebSocket.js (973 bytes)
HTTP > request: /main.css
HTTP > file sent: /main.css (988 bytes)
HTTP > request: /function.js
HTTP > file sent: /function.js (943 bytes)
```
so it looks like files are requests are getting through and files are being served..

I am seeing ERR_CONTENT_LENGTH_MISMATCH errors in the console for main.css and function.js but I am not certain this actually prevents them downloading?

Any thoughts on what this might be?   I have tried in Chrome and Firefox and both show the same issue

**Comment by visualapproach** [comment]
I encountered that too while developing v4. From the error message, I guess it's when the ESP doesn't deliver the whole file, or maybe taking too long. If it's something that's changed in the browser, the ESP wifi libraries or memory shortage, I don't know. But it works again for me. I've had one occasion but I waited a while and retried and it worked.

**Reply by visualapproach** [reply]
what firmware are you using?

**Reply by jamesonuk** [reply]
I am using the 6 wire version.

It is a bit odd as it looks like the HTTP response above said it was 988 bytes but the file is around 7.2k.   Trying to get the css via curl I get
```
HTTP > request: /main.css
HTTP > file sent: /main.css (7328 bytes)
```
which is the right size

**Reply by visualapproach** [reply]
Hi, I ment what firmware version. 3.x or 4? V4 has date 2023-02-05-alpha

**Reply by visualapproach** [reply]
![image](https://user-images.githubusercontent.com/22987250/216844592-3392a2e0-2900-4a80-9f92-614d09df755a.png)

**Reply by taferris** [reply]
How is your wifi connection?   Take a look at the RSSI value on main page.

I see this type of issue occasionally on my setup.  My RSSI is barely above the noise floor and so I have a very poor connection - worse with bad weather.  I have been thinking about changing the web socket update from every 2 seconds to every 10 seconds, so that it has plenty of time for retransmits.

**Reply by jamesonuk** [reply]
As I am just testing it is sat pretty close to the router with an RSSI of -40  so if anything the signal might be too strong

**Reply by jamesonuk** [reply]
> Hi, I ment what firmware version. 3.x or 4? V4 has date 2023-02-05-alpha

3.5.1

**Reply by jamesonuk** [reply]
> ![image](https://user-images.githubusercontent.com/22987250/216844592-3392a2e0-2900-4a80-9f92-614d09df755a.png)

Possibly unnrelated or a red herring but I did see a 404 for websoxket.js at one point (note it was looking for all lower case where as file name is in camel case.   I had a quick look and couldn't see where this would have come from thiugh

**Comment by jamesonuk** [comment]
I had been playing around with a few things without much luck but I minified the CSS and function.js files and now it seems to work fine.
I have also renamed websockets.js to lowercase (and changed the references I found) but I think that is completely unrelated

**Comment by visualapproach** [comment]
This started happening both to v3.5 and v4 so there must be something that changed. Can someone try downgrading some libraries. Starting with
platform = espressif8266@~3.0.0 or even ~2.0.0
Thx

**Reply by jamesonuk** [reply]
Did some playing with a few libraries earlier but without much luck.   Changing the platform to ~3.0.0 seemed to give the same error and ~2.0.0 (and ~2.6.3 for that matter) gave compilation errors I didn't have time to delve into.

I flashed a brand new board (albeit the same model as the one I had issues with) and the same issue there.

I will try some more combinations and try and see if I can get back to a working version

**Reply by jamesonuk** [reply]
going the other way 
```
platform = espressif8266@~4.1.0
```
however seems to currently be stable (but I had it all working nicely this morning after minimising the files and then it wen back later...)

**Reply by visualapproach** [reply]
Ok good to know. I think there is not enough memory to send large messages. Just a hunch. I don't know if it's dram or sram(iram) but you could try adding a flag in platformio.ini that increases the iram to 48K. See how I did that in development branch. It's a shot in the dark though. Appreciate your help with testing

**Reply by jamesonuk** [reply]
I don't know enough about the web server implementation here but I do a lot with full size web servers at work and I can't see memory being the issue.  

It should be sending the data in chunks but it seems as if the server is sending the response before it is ready and  it is saying the file is smaller than it is and then you get the error.

Just looking now and the hwtest page is still problematic with the image showing the problem but the other pages are fine.

Going to do a bit  more digging

**Reply by eried** [reply]
> going the other way
> 
> ```
> platform = espressif8266@~4.1.0
> ```
> 
> however seems to currently be stable (but I had it all working nicely this morning after minimising the files and then it wen back later...)

This seems to have fixed my issue (with latest december release)

**Reply by jamesonuk** [reply]
> Just a hunch. I don't know if it's dram or sram(iram) but you could try adding a flag in platformio.ini that increases the iram to 48K

added `-DPIO_FRAMEWORK_ARDUINO_MMU_CACHE16_IRAM48` and still seeing the same as above.   All working except for the images on the hwtest page which still show the `ERR_CONTENT_LENGTH_MISMATCH` error.

It is obviously related to file size and it is looking more likely that it is a feature of the ESP8266WebServer.
It has been a long time since I had to do any C++ at University and microcontroller stuff is all new to me but I came across a couple of links that might make sense to you....
https://www.esp8266.com/viewtopic.php?f=32&t=10189
https://www.esp8266.com/viewtopic.php?f=32&t=14267

There is also https://github.com/me-no-dev/ESPAsyncWebServer which is built on top of ESPAsyncTCP which you are already using

**Reply by visualapproach** [reply]
Yes, I've read those links and there are no real answer I think to other ppl with the same problem. The same code worked fine awhile ago so I think something has changed in either the 8266webserver library, or the arduino framework or espressif8266 core. Someone tried downgrading the core with no difference, so probably not that. I'll investigate further when I have time.
Regarding the ESPAsyncWebServer, I ported the code to try it out. It took a whole day but I read good stuff about it. But it ended up making the .bin file and RAM usage 50% bigger* and I still experienced the broken html pages. So I switched back to "native" webserver and for some reason it just works for me. I have the exact same v4 as is published here on github. If I open three simultaneous web pages it reboots the ESP but that is more understandable.
*I had to install other libraries also to make it work.

**Comment by eried** [comment]
mine continues failing from time to time, no idea why :( I wish this was ported this to esp32

**Reply by visualapproach** [reply]
Your comment is not very much to go forward on, but maybe you just wanted to let some steam off :-)
If you have a 6-wire pump, try v4, and remember to build+upload the data files as well since they are compacted and changed. If you want to save any settings files, do so prior to uploading the data files.

**Reply by eried** [reply]
yeah I know I was just adding info to my previous "success" :D but v4.1.0 actually did not solved it. 

What other version are you using for best stability?

**Reply by visualapproach** [reply]
I don't "use" any version since I don't have a spa anymore. I have a pump acting as a test-bed. I claim that v3.5.1 is stable in itself. Something else have changed. I'd say both 3.5.1 and 4.0.0 is good as far as web serving is concerned for me. I have a PRE2021, static IP, running 4.0.0 from today. Updated VSC, platformio, all libraries, esp8266 and arduino framework.

**Reply by eried** [reply]
OK so I was testing in a duplicated esp, and started doing the same, it is a bootloop after is configured, sometimes I reset my router and HA and it goes for few hours or days but then one day starts bootlooping again. The clone shows the same behaviour (after mqtt and wifi is configured):

```
WiFi > Connected.
 SSID: "Skynet"
 IP: "192.168.212.80"
OTA > ready
HTTP > server started
WebSocket > server started
MQTT > Buffer size successfully increased
End of setup()

 ets Jan  8 2013,rst cause:2, boot mode:(3,6)

load 0x4010f000, len 3424, room 16
tail 0
chksum 0x2e
load 0x3fff20b8, len 40, room 8
tail 0
chksum 0x2b
csum 0x2b
v0007c030
~ld
WiFi > using WiFi configuration with SSID "Skynet"
```

I commented out all the MQTT stuff and the bootloop continues, so it is something else. Any hint what can produce this weird behaviour? 

I tried with the latest of all the libraries too, and it is the same.

**Reply by eried** [reply]
OK. actually I think I found the issue... which is strange. If you initialize the commandQ:

![image](https://user-images.githubusercontent.com/1091420/218321699-f65ac9c3-bc6b-47fb-a011-2b9f87480130.png)

There is no reboot. You think it might be possible that "noise" can generate this state of the esp rebooting all the time? :O

**Reply by visualapproach** [reply]
Unititialized objects can be dangerous so possibly that helps. I had an occasion of reset command being saved in cmdq.txt which was !fun.
Let us know how it turns out after a while.

**Reply by eried** [reply]
It is a beast now. It makes sense now that it was so fragile before... Because I had a daily reboot to try to prevent this instability but that was generating it... The boot process having this randomness

**Reply by SchallertDK** [reply]
Has the solution been found? 
Couldn't read/find the "answer" in the feed (?)
I have same reboot issues always with this "code":
![image](https://user-images.githubusercontent.com/103250760/220064290-632c53f2-ef1e-431d-abfd-13d311a644a3.png)

I have made some automations for start heating the spa when the kWh price is low, and stop the heating when the prise i high. Working fine, but when the reboot has ben, the:
![image](https://user-images.githubusercontent.com/103250760/220064864-b2f28b75-913f-49f7-a1d1-86f14bea52f5.png)
is back to 20 C.
And i cant figure where and how i can ask in the automation, to put back on 38 C. I can only do that manuel at the lovelace card:
![image](https://user-images.githubusercontent.com/103250760/220065434-b2ce86b5-7bcb-4d3b-b8fb-cc94f8bb7102.png)

**Reply by eried** [reply]
I need to correct my previous comment :/ it is still a bit unstable, I think is MQTT in my case. I will research more later, from time to time I just need to reboot the board (new esp and all) and reset wifi credentials and such

**Reply by visualapproach** [reply]
> Has the solution been found? Couldn't read/find the "answer" in the feed (?) I have same reboot issues always with this "code": ![image](https://user-images.githubusercontent.com/103250760/220064290-632c53f2-ef1e-431d-abfd-13d311a644a3.png)

Not sure. May I suggest you try version 4 beta. Chances are that it behaves better since it set the clockspeed to 160MHz and gzip the data files. Also, it doesn't like too many open connections, like 4 webpages running at the same time. Or sending messages super frequently, like several times per second. I think I have my MQTT set to every 60 seconds.

> I have made some automations for start heating the spa when the kWh price is low, and stop the heating when the prise i high. Working fine, but when the reboot has ben, the: ![image](https://user-images.githubusercontent.com/103250760/220064864-b2f28b75-913f-49f7-a1d1-86f14bea52f5.png) is back to 20 C. And i cant figure where and how i can ask in the automation, to put back on 38 C. I can only do that manuel at the lovelace card:

I will probably save the target temerature and load it again if the "restore states" flag is set. In upcoming updates. But you can send the messages from MQTT to change temperature: 
```
service: mqtt.publish
data:
  topic: layzspa/command
  payload: '{"CMD":0,"VALUE":38,"XTIME":0,"INTERVAL":0,"TXT":"hello"}'
```
Remove the TXT parameter if you're on v3.

**Reply by SchallertDK** [reply]
I cant see how to update to development v4. Last year i could, but this update i cant see how to those pump and pcb version like last time you helped me with my firmware "4w_2022_07_21"
If i read the guide i cant see how it should be done 😅

**Reply by visualapproach** [reply]
@SchallertDK see point 11 in the readme. You select it from web UI

**Reply by SchallertDK** [reply]
And how do i get the downloadet zip file to the esp module? By the "File Uploader" at the web UI? And if, which file do i have to upload?

**Reply by visualapproach** [reply]
As in #342 and the build instructions. The build instructions is not updated for v4 but uploading is the same except now you don't need to edit the code. Compile and upload the code, then build and upload the file system. After that you visit all the pages from the hamburger menu in the web UI and set up things. The saved statistics like uptime etc will be reset. If you don't want that you need to upload every single file from the data/ directory by using the file uploader.

**Reply by SchallertDK** [reply]
When i open Visual Studio, entering PlatfromIO, open project, and select the "code" folder.
And Push the arrow in the buttom "PlatfomIO: Upload" - This error enter (?)
![image](https://user-images.githubusercontent.com/103250760/221601820-134940c4-0376-4a1d-88df-f6caa6d40e5b.png)

**Reply by visualapproach** [reply]
Clean all from platformio menu and try again

**Reply by SchallertDK** [reply]
Still same error.
I have removed folder from workspace, and then selected the folder with the platformio.ini file
But at the PlatformIO there is nothing? It only shows after a correct PlatformIO upload, right?
![image](https://user-images.githubusercontent.com/103250760/221614295-c24d78b0-0136-467e-9dce-f3eb676c072f.png)

**Reply by SchallertDK** [reply]
Do i have to umount the ESP from the spa and cable connect it to my pc? Or is it still posible to du this update OTA? (Is this maybe why i got the error?)

**Reply by visualapproach** [reply]
> Do i have to umount the ESP from the spa and cable connect it to my pc? Or is it still posible to du this update OTA? (Is this maybe why i got the error?)

No you can upload OTA. Only if it fails to reboot properly do you need to use cable.
Make sure you have updated platformio and VSC. Click the platformio icon and choose open project. Open the code folder. Compile + upload. Build filesystem image + upload (choose the one without OTA. It will do OTA anyway)

**Reply by SchallertDK** [reply]
Where is the Compile buttom? I see the Upload buttom but not any compile

**Reply by SchallertDK** [reply]
Where is the "Compile" button? :)

**Reply by visualapproach** [reply]
Compile and build is the same thing and page 10-11 in the build instructions shows how to upload to esp. You can also read about it in the platformio docs.

**Reply by SchallertDK** [reply]
All the "Hardware Watchdog" faults is only appearing when the heating is getting over 31-32 degrees? My setpoint is 38.
I dont know if that can help anything in the troubleshooting? When i heat up fra example 10 degrees and all the way up to this 31-32 degrees, there is no faults, but efter that i have fault sometimes every 10 minutes

**Reply by visualapproach** [reply]
@SchallertDK that's strange. If it actually has to do with the temperature maybe a bad solder joint is expanding a little giving bad connection. Or maybe it's not related to the actual temperature but the running time, which coincide with heating time to 32 degrees. Like if there is a memory leak accumulating over time. Can't advice anything else for the moment than to try the latest beta version as mentioned at the bottom of this page.

**Comment by eried** [comment]
🧐 in still trying to find out why mqtt generates issues

![Screenshot_2023-02-22-18-01-40-838_io homeassistant companion android](https://user-images.githubusercontent.com/1091420/220702566-d5b0e2e0-8fde-44f9-8b37-8227cf55498e.jpg)

There is some spotty connections 😪 but when I operate the layzspa manually it responds ok... Misterious

**Reply by visualapproach** [reply]
Don't know. Check RSSI graph if the wi-fi is bad on those occasions. Maybe using a static IP will help a little by connecting faster after a drop.

**Comment by visualapproach** [comment]
I've optimized the code to use less RAM and in my testing that got rid of the error mentioned in OP.

khoih-prog:
RAM:   [======    ]  58.8% (used 48164 bytes from 81920 bytes)
Flash: [=======   ]  66.5% (used 694928 bytes from 1044464 bytes)

replace khoih with tzapu:
RAM:   [======    ]  57.0% (used 46680 bytes from 81920 bytes)
Flash: [======    ]  63.0% (used 657684 bytes from 1044464 bytes)

Saved 1484 bytes RAM, 37244 bytes Flash

removed double definitions of CHARS[]:
RAM:   [======    ]  56.9% (used 46572 bytes from 81920 bytes)
Flash: [======    ]  63.0% (used 657688 bytes from 1044464 bytes)

Saved another 108 bytes of RAM

moved strings to Flash:
RAM:   [======    ]  56.2% (used 46076 bytes from 81920 bytes)
Flash: [======    ]  63.0% (used 657684 bytes from 1044464 bytes)

Saved another 496 bytes of RAM

Move display text string outside sStates into DSP object:
RAM:   [======    ]  56.1% (used 45948 bytes from 81920 bytes)
Flash: [======    ]  62.8% (used 656168 bytes from 1044464 bytes)

Saved another 128 bytes of RAM
Total: 2216 bytes of RAM (~10% more free DRAMmemory) and 38760 bytes of flash memory.

Present code is in dev branch 2023-03-25-1526
